### ISSUE TITLE
[feature request] Simplified gather with not specifying full index

### ISSUE THREAD

**Issue by vadimkantorov** [pytorch/pytorch#64208]
E.g. we have `I = torch.randint(0, n3, (n1, n2))` and `T = torch.rand(n1, n2, n3, n4, n5)`
We'd like to compute `O[i, j, ...] = T[i, j, I[i, j], ...]`

This is fairly frequent. Currently this is possible with a combination of multi unsqueeze + expand + gather.

I propose a new helper function to do this in one ago or just making gather support this directly by relaxing index tensor shape constraints

**Comment by vadimkantorov** [comment]
Real-world example: https://github.com/pytorch/vision/blob/7947fc8fb38b1d3a2aca03f22a2e6a3caa63f2a0/torchvision/models/detection/roi_heads.py#L129

Here it is for 1D index, then this functionality is possible by using advanced indexing with two tensors. But for larger dimensions this is not possible. Also, this particular form of advanced indexing is not very well-known.

**Comment by vadimkantorov** [comment]
My implementation:
```python
gather_incomplete = lambda tensor, I: tensor.gather(I.ndim, I[(...,) + (None,) * (tensor.ndim - I.ndim)].expand((-1,) * (I.ndim + 1) + tensor.shape[I.ndim + 1:])).squeeze(I.ndim)
```

In addition to being quite convoluted, this would also quite likely fail in JIT

**Comment by vadimkantorov** [comment]
This could also be used for generalizing F.cross_entropy accepted batch shapes

**Comment by vadimkantorov** [comment]
Alternatively, index_select could be extended somehow to allow >1dim index

**Comment by zou3519** [comment]
This is an interesting ask. Are there other libraries that have something like this?

**Comment by vadimkantorov** [comment]
Never saw anything like this in libraries, but in practice this generalization of index_select / simplification of tather is very frequent

**Comment by vadimkantorov** [comment]
So one could think of this as batched index_select

**Comment by vadimkantorov** [comment]
cc @Chillee

**Comment by julywater** [comment]
I am also looking for that function

**Comment by vadimkantorov** [comment]
A similar function is also needed for permuting / batch-indexing across a dimension, e.g.
``` 
(Pdb) I.shape
torch.Size([4, 4300])
(Pdb) x.shape
torch.Size([4, 4300, 256])
(Pdb) x.gather(1, I.unsqueeze(-1).expand_as(x)).shape
torch.Size([4, 4300, 256])
```

where I is permutation / batched index across a dimension

**Comment by vadimkantorov** [comment]
Is https://pytorch.org/docs/stable/generated/torch.take_along_dim.html?highlight=take_along_dim#torch.take_along_dim related to this request?

**Comment by Goose-Bomb** [comment]
> This is an interesting ask. Are there other libraries that have something like this?

It seems that [tf.gather](https://www.tensorflow.org/api_docs/python/tf/gather) has a `batch_dims` parameter that coud achieve the batched index selection.

For example, when `batch_dims=1, axis=1`, it can achieve:

```
output[i, j, ...] = input[i, indices[i, j], ...]
```

**Comment by sanketkshah** [comment]
Also looking for such a function!

**Comment by jinfagang** [comment]
Am also looking for a way to more simplied to do this:

```
mask: [3, 100, 224, 224]
keep: [3, 20]

expect: [3, 20, 224, 224]
```

Currently can not find a way without loop to do it.

**Comment by vadimkantorov** [comment]
@jinfagang I think https://github.com/pytorch/pytorch/issues/64208#issuecomment-909077669 this should help, but its code is quite cumbersome

**Comment by gorold** [comment]
Does ```vmap``` solve this issue? Something like this:

```python
input = torch.randn(32, 10, 3)
index = torch.randint(0, 10, size=(32, 2))

indexed_input = vmap(
    torch.index_select,
    in_dims=(0, None, 0),
)(input, 0, index)
```

**Comment by vadimkantorov** [comment]
several aspects:
1) i don't know if unsqueeze+expand+gather is as performant or not and if any of these ways is optimal
2) it would be best to have a fully dynamic dimension version by default, e.g. that if you index on the left, you don't even need to specify any dimensions at all
3) it should be exportable/scriptable etc - for that maybe supporting a special op is best?
4) even the proposed workaround are optimal, the operation IMO is still frequent and it would be best to include it in core (even if it's implemented in terms of other existing functions) or extend existing `index_select` to support batched mode

**Comment by zou3519** [comment]
Hi-pri for user activity. We have a couple of options here:
- extend index_select
- make a new operator
- use vmap(index_select) as the API

**Comment by vadimkantorov** [comment]
Related: https://github.com/pytorch/pytorch/issues/9407

**Comment by vadimkantorov** [comment]
Also, might be an interesting idea if einsum-like named expressions were supported, so that something like `O[i, j, ...] = T[i, j, I[i, j], ...]` or `O[b, i, k] = T[b, I[i, k]]` could be directly used

**Comment by K-Nick** [comment]
Your updated description is clear and provides a good explanation of the functionality of your einops-style generalized gather operation. Here's a slightly polished version to ensure maximum clarity and professionalism:

---

I've implemented an einops-style generalized gather operation that has passed randomized tests with perfect results. Below is part of the rule I used for parsing the `pattern`:

$$
    \text{m n[m k l] d -> m k l d}
$$

This transformation results in $\text{output[m, k, l, :] = input[m, indices[m, k, l], :]}$.

To further clarify, here are some key differences from `torch.gather`:

- The positions of axes are flexible on both sides. For example, in the original `torch.gather`, dimensions are fixed when `dim=2` and `ndim=3`, which leads to `output[i][j][k] = input[i][j][indices[i][j][k]]`.
- Multiple batch axes in `indices` are supported (e.g., K in this case).
- `Indices` can ignore certain axes from the tensor (e.g., D in this case). These axes will expand to form a hyperplane, meaning we are gathering a "hyperplane" instead of a singular "value".
- The output shape can be rearranged in an einops-style, thanks to powerful `rearrange`.

**Comment by grofte** [comment]
I've used Keras + Tensorflow for the last 5-6 years but I'm new to PyTorch. So far Keras with a PyTorch backend is a bit chaotic but it works fine. What I would do before was a top-k and then run the indices through gather with the labels and everything would get mapped back. No carrying around the labels in a separate artifact and mapping them afterwards. I think I can bundle it all up if I put my model into a HuggingFace pipeline?

Anyway, for the equivalent here I have to do this 

```python
    def call(self, inputs):
        values, indices = torch.topk(inputs, self.k)
        labels = [torch.index_select(self.labels, 0, idx) for idx in indices]
        return values, labels
```

and this only works because I can express my labels as integers (I will still have to zfill them afterwards since they really are 8 digit strings). I seems janky and now my labels are a list of rank-one arrays instead of a rank-two array like my softmax values.

**Comment by the-jb** [comment]
Use [torch.take_along_dim](https://pytorch.org/docs/stable/generated/torch.take_along_dim.html).

```python
import torch

inputs = torch.randn(B, L, K, E)  # size = (B, L, K, E)
indices = torch.randint(0, K, size=(B, 1, k, 1)) # size = (B, 1, k, 1) where k <= K

out = torch.take_along_dim(inputs, dim=2, indices=indices)  # size = (B, L, k, E)
```
